src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -0,0 +1,199 @@
+    enqueue_locked(m);
+  }
+}
+
+AsyncLogWriter::AsyncLogWriter()
+  : _state(ThreadState::NotReady),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    _state = ThreadState::Initialized;
+  }
+
+  log_info(logging)("The maximum entries of AsyncLogBuffer: " SIZE_FORMAT ", estimated memory use: " SIZE_FORMAT " bytes",
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+class AsyncLogMapIterator {
+  AsyncLogBuffer& _logs;
+
+ public:
+  AsyncLogMapIterator(AsyncLogBuffer& logs) :_logs(logs) {}
+  bool do_entry(LogFileOutput* output, uint32_t* counter) {
+    using none = LogTagSetMapping<LogTag::__NO_TAG>;
+
+    if (*counter > 0) {
+      LogDecorations decorations(LogLevel::Warning, none::tagset(), output->decorators());
+      stringStream ss;
+      ss.print(UINT32_FORMAT_W(6) " messages dropped due to async logging", *counter);
+      AsyncLogMessage msg(*output, decorations, ss.as_string(true /*c_heap*/));
+      _logs.push_back(msg);
+      *counter = 0;
+    }
+
+    return true;
+  }
+};
+
+void AsyncLogWriter::perform_IO() {
+  // use kind of copy-and-swap idiom here.
+  // Empty 'logs' 'swaps' the content with _buffer.
+  // Along with logs destruction, all procceeded messages are deleted.
+  //
+  // the atomic operation 'move' is done in O(1). All I/O jobs are done without lock.
+  // This guarantees I/O jobs don't block logsites.
+  AsyncLogBuffer logs;
+  { // critical region
+    AsyncLogLocker ml;
+    AsyncLogMapIterator dropped_counters_iter(logs);
+
+    _buffer.pop_all(&logs);
+    // append meta-message of dropped counters
+    _stats.iterate(&dropped_counters_iter);
+  }
+
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    char* msg = e->message();
+
+    if (msg != nullptr) {
+      e->output()->write_blocking(e->decorations(), msg);
+      os::free(msg);
+    }
+  }
+}
+
+void AsyncLogWriter::run() {
+  assert(_state == ThreadState::Running, "sanity check");
+
+  while (_state == ThreadState::Running) {
+    _sem.wait();
+    perform_IO();
+  }
+
+  assert(_state == ThreadState::Terminated, "sanity check");
+  perform_IO(); // in case there are some messages left
+}
+
+AsyncLogWriter* AsyncLogWriter::_instance = nullptr;
+
+void AsyncLogWriter::initialize() {
+  if (!LogConfiguration::is_async_mode()) return;
+
+  if (_instance == nullptr) {
+    AsyncLogWriter* self = new AsyncLogWriter();
+
+    if (self->_state == ThreadState::Initialized) {
+      Atomic::release_store_fence(&AsyncLogWriter::_instance, self);
+      // all readers of _instance after fence see NULL,
+      // but we still need to ensure no active reader of any tagset.
+      // After then, we start AsyncLog Thread and it exclusively takes
+      // over all logging I/O.
+      for (LogTagSet* ts = LogTagSet::first(); ts != NULL; ts = ts->next()) {
+        ts->wait_until_no_readers();
```

Comment:
I would like to consolidate "exclusively". Once async logging takes over, we guarantee that no synchronous logsite is writing files in concurrency.  If we guarantee that, we can avoid FileLocker() because only AsyncLog Thread writes files. 

Fence isn't enough here. fence can only guarantee that logsites 'after' the fence use async logging.  It's possible that there are on-going synchronous logging. 

This is a RCU trick of Unified logging.  RCU is kind of synchronization biased to multiple readers. `LogOutputList::Iterator` actually embeds an atomic counter.
ts->wait_until_no_readers() waits until all readers finish. It means that all synchronous write have done. 

here is how synchronous log writes. Iterator increases to zero when the loop is done. 
 
```cpp
void LogTagSet::log(LogLevelType level, const char* msg) {
  LogDecorations decorations(level, *this, _decorators);
  for (LogOutputList::Iterator it = _output_list.iterator(level); it != _output_list.end(); it++) {
    (*it)->write(decorations, msg);
  }
}```

Reply:
Thanks for the explanation. I assume you meant that the `Iterator` **de**creases the readers to zero when the loop is done. But I think I got it now.

Reply:
Yes, the scope of 'it' is defined inside the 'for' construct.  Once control flow exists the loop, 
the dtor of Iterator will automatically be invoked. When all pending synchronous logsites have completed,  all RCU counters of OutputList objects restore to zero.  It means there's no on-going 'reader'. 
 

```
    ~Iterator() {
      _list->decrease_readers();
    }
```